src/psyclone/tests/domain/lfric/transformations/lfric_extract_test.py

```diff
@@ -0,0 +1,698 @@
+      !
+      CALL psy_data%PreStart("single_invoke_psy", """ \
+      """"invoke_0:inc_ax_plus_y:r0", 0, 1)
+      CALL psy_data%PreDeclareVariable("df_post", df)
+      CALL psy_data%PreEndDeclaration
+      CALL psy_data%PreEnd
+      !$omp parallel do default(shared), private(df), schedule(static)
+      DO df=1,undf_any_space_1_f1
+        f1_proxy%data(df) = 0.5*f1_proxy%data(df) + f2_proxy%data(df)
+      END DO
+      !$omp end parallel do
+      CALL psy_data%PostStart
+      CALL psy_data%ProvideVariable("df_post", df)
+      CALL psy_data%PostEnd
+      !
+      ! ExtractEnd"""
+    assert output in code_omp
+
+    # TODO #646
+    # At this stage not all required parameters are passed via PSyData.
+    # This is an except of missing lines, which will cause this test to x-fail
+    not_yet_working = ['CALL psy_data%ProvideVariable("f1", f1)',
+                       'CALL psy_data%ProvideVariable("f2", f2)']
+    for line in not_yet_working:
+        if line not in code:
+            pytest.xfail("#646 LFRic extraction not fully supported yet.")
+        if line not in code_omp:
+            pytest.xfail("#646 LFRic extraction not fully supported yet.")
+    assert False, "X-failing test suddenly working: #646 lfric extraction."
+
+
+def test_extract_kernel_and_builtin_dynamo0p3(tmpdir):
+    ''' Test that extraction of a Kernel and a BuiltIny in an Invoke
+    produces the correct result in Dynamo0.3 API. '''
+    etrans = LFRicExtractTrans()
+
+    psy, invoke = get_invoke("15.1.2_builtin_and_normal_kernel_invoke.f90",
+                             DYNAMO_API, idx=0, dist_mem=False)
+    schedule = invoke.schedule
+
+    schedule, _ = etrans.apply(schedule.children[1:3])
+    code = str(psy.gen)
+    output = """
+      ! ExtractStart
+      !
+      CALL psy_data%PreStart("single_invoke_builtin_then_kernel_psy", """ \
+      """"invoke_0:r0", 1, 3)
+      CALL psy_data%PreDeclareVariable("f2", f2)
+      CALL psy_data%PreDeclareVariable("cell_post", cell)
+      CALL psy_data%PreDeclareVariable("df_post", df)
+      CALL psy_data%PreDeclareVariable("f3_post", f3)
+      CALL psy_data%PreEndDeclaration
+      CALL psy_data%ProvideVariable("f2", f2)
+      CALL psy_data%PreEnd
+      DO df=1,undf_any_space_1_f2
+        f2_proxy%data(df) = 0.0
+      END DO
+      DO cell=1,f3_proxy%vspace%get_ncell()
+        !
+        CALL testkern_code_w2_only(nlayers, f3_proxy%data, """ + \
+        """f2_proxy%data, ndf_w2, undf_w2, map_w2(:,cell))
+      END DO
+      CALL psy_data%PostStart
+      CALL psy_data%ProvideVariable("cell_post", cell)
+      CALL psy_data%ProvideVariable("df_post", df)
+      CALL psy_data%ProvideVariable("f3_post", f3)
+      CALL psy_data%PostEnd
+      !
+      ! ExtractEnd"""
+
+    assert output in code
+
+    assert LFRicBuild(tmpdir).code_compiles(psy)
+
+    # TODO #646 (LFRic not fully supported) and #637 (no builtin support)
+    # This is an except of missing lines, which will cause this test to x-fail
```

Comment:
s/except/excerpt/

Reply:
Only problem ... I don't know how ... but I am sure I can get some investors for the idea to solve C&P problems.

Reply:
I'm sure the solution lies in machine learning and AI. You'd definitely get investment for that.